tools/webtransport/h3/webtransport_h3_server.py

```diff
@@ -121,47 +125,50 @@ def _h3_event_received(self, event: H3Event) -> None:
                     data=event.data,
                     stream_ended=event.stream_ended)
             elif isinstance(event, DatagramReceived):
-                self._handler.datagram_received(data=event.data)
+                if self._allow_datagrams:
+                    self._handler.datagram_received(data=event.data)
 
     def _receive_data_on_session_stream(self, data: bytes, fin: bool) -> None:
         self._capsule_decoder_for_session_stream.append(data)
         if fin:
             self._capsule_decoder_for_session_stream.final()
         for capsule in self._capsule_decoder_for_session_stream:
-            if capsule.type == CapsuleType.DATAGRAM:
-                raise ProtocolError(
-                    "Unimplemented capsule type: {}".format(capsule.type))
-            if capsule.type == CapsuleType.REGISTER_DATAGRAM_CONTEXT:
+            if capsule.type in {CapsuleType.DATAGRAM,
+                                CapsuleType.REGISTER_DATAGRAM_CONTEXT,
+                                CapsuleType.CLOSE_DATAGRAM_CONTEXT}:
                 raise ProtocolError(
                     "Unimplemented capsule type: {}".format(capsule.type))
-            elif capsule.type == CapsuleType.REGISTER_DATAGRAM_NO_CONTEXT:
-                # TODO(yutakahirano): Check the Datagram Format Type.
-                # TODO(yutakahirano): Check that this arrives before any
-                # datagrams/streams requests.
-                if self._close_info is not None:
-                    raise ProtocolError(
-                        "REGISTER_DATAGRAM_NO_CONTEXT after " +
-                        "CLOSE_WEBTRANSPORT_SESSION")
-            elif capsule.type == CapsuleType.CLOSE_DATAGRAM_CONTEXT:
-                raise ProtocolError(
-                    "Unimplemented capsule type: {}".format(capsule.type))
-            elif capsule.type == CapsuleType.CLOSE_WEBTRANSPORT_SESSION:
-                if self._close_info is not None:
-                    raise ProtocolError(
-                        "CLOSE_WEBTRANSPORT_SESSION arrives twice")
+            if capsule.type in {CapsuleType.REGISTER_DATAGRAM_NO_CONTEXT,
+                                CapsuleType.CLOSE_WEBTRANSPORT_SESSION}:
+                # We'll handle this case below.
+                pass
             else:
                 # We should ignore unknown capsules.
                 continue
 
+            if self._close_info is not None:
+                raise ProtocolError((
+                    "Receiving a capsule with type = {} after receiving " +
+                    "CLOSE_WEBTRANSPORT_SESSION").format(capsule.type))
+
+            if capsule.type == CapsuleType.REGISTER_DATAGRAM_NO_CONTEXT:
+                buffer = Buffer(data=capsule.data)
+                format_type = buffer.pull_uint_var()
+                WEBTRANPORT_FORMAT_TYPE = 0xff7c00
```

Comment:
Please add a link to the spec, and maybe put this in somewhere else (e.g. toplevel of this file around the beginning)

Reply:
Added a link to the spec.